lib/chipmunk/incoming_storage.rb

```diff
@@ -1,12 +1,46 @@
 # frozen_string_literal: true
 
+require "pathname"
+
 module Chipmunk
   # IncomingStorage is a factory for proxies to storage of incoming deposits.
   class IncomingStorage
 
+    class UserPathBuilder
+      def initialize(root_path)
+        @root_path = Pathname.new(root_path)
+        raise ArgumentError, "root_path must not be nil" unless @root_path
+      end
+
+      def for(package)
+        File.join(root_path, package.user.username, package.bag_id)
+      end
+
+      private
+
+      attr_reader :root_path
+    end
+
+    class IdPathBuilder
+      def initialize(root_path)
+        @root_path = Pathname.new(root_path)
+        raise ArgumentError, "root_path must not be nil" unless @root_path
+      end
+
+      def for(package)
+        File.join(root_path, package.bag_id)
+      end
+
+      private
+
+      attr_reader :root_path
+    end
+
     # Create an IncomingStorage instance.
-    def initialize(volume:)
+    def initialize(volume:, paths:, links: )
```

Comment:
These names are not intuitive to me -- it's hard to figure what paths and links mean here. I guess this is probably flowing downhill from "upload_link". In that sense, I'm willing to deal with this for now, despite it not really being a link.

Let me try to summarize:

```
# @param volume [Volume] The Volume from which the deposited packages should be ingested
# @param paths [PathBuilder] A PathBuilder that yields a path on disk to the user upload location for a deposit, based on a package
# @param links [PathBuilder] A PathBuilder that yields an rsync destination to which the user should upload, for a given package
```